File: GREEN-DB_query.py
'''
RegDB query
Retrieve detailed information from RegDB for regions/vars of interes
Input can be
- list of region IDs
- a VCF annotated with GREEN-VARAN
- a tables containing variant ID (chr_pos_ref_alt), comma-separated regionIDs 
Author: Edoardo Giacopuzzi
'''

import argparse, os, re, sys, subprocess, time
from datetime import datetime,timedelta
import sqlite3
import logging
import pandas as pd
from cyvcf2 import VCF

VERSION=1.0
pd.set_option('mode.chained_assignment',None)
BASE_DIR = os.path.dirname(os.path.realpath(sys.argv[0]))
RESOURCE_DIR = "resources"
REG_DB="SQlite/RegulatoryRegions.db"
DB_FILE = "{base_dir}/{resource_dir}/{reg_db}".format(base_dir=BASE_DIR,resource_dir=RESOURCE_DIR,reg_db=REG_DB)
TABLES = ["regions","gene_details","pheno_details", "DNase","dbSuper","TFBS","UCNE"]
TAB_HEADERS = {
    'regions' : ['regionID','chrom','start','stop','type','std_type','DB_source','PhyloP100_median','constraint_pct','controlled_gene','closestGene_symbol', 'closestGene_ensg', 'closestGene_dist','cell_or_tissues','detection_method','phenotype'],
    'gene_details' : ['regionID','chrom','start','stop','std_type','controlled_gene','detection_method','tissue_of_interaction'],
    'pheno_details' : ['regionID','chrom','start','stop','std_type','phenotype','detection_method','DB_source'],
    'DNase' : ['regionID','DNase_chrom','DNase_start','DNase_stop','DNase_ID','DNase_cell_or_tissue'],
    'dbSuper' : ['regionID','dbSuper_chrom','dbSuper_start','dbSuper_stop','dbSuper_ID','dbSuper_cell_or_tissue'],
    'TFBS' : ['regionID','TFBS_chrom','TFBS_start','TFBS_stop','TF_name','TFBS_cell_or_tissue'],
    'UCNE' : ['regionID','UCNE_chrom','UCNE_start','UCNE_stop','UCNE_ID']
}

def exitNow(start_time):
    end_time = time.time() - start_time
    end_time = str(timedelta(seconds=round(end_time)))
    logger.info("Process completed in %s", end_time)
    sys.exit()

#create string with current time (short) or date+time(long)
def now(sep=":", string_format="short"):
    now = datetime.now()
    if string_format == "short": 
        current_time = now.strftime("%H{sep}%M{sep}%S".format(sep=sep))
    elif string_format == "long":
        current_time = now.strftime("%Y%m%d_%H{sep}%M{sep}%S".format(sep=sep)) 
    return current_time

#Read genes list, either comma-separated or file with list
def readList(ID_list):
    if os.path.isfile(ID_list):
        gene_list = open(ID_list).read().splitlines()
    else:
        gene_list = ID_list.split(",")
    return(gene_list)

#create a database connection to the SQLite database
def create_connection(db_file):
    conn = None
    try:
        conn = sqlite3.Connection(db_file)
        cur = conn.cursor()
        cur.execute('PRAGMA cache_size = 1000000')
        cur.execute('PRAGMA locking_mode = EXCLUSIVE')
        cur.execute('PRAGMA temp_store = 2')
        cur.execute('PRAGMA synchronous = 0')
        cur.execute('PRAGMA journal_mode = OFF')
        cur.execute('PRAGMA threads = 8')
    except:
        conn = None
 
    return conn

#prepare query for methods or genes tables 
def prepare_query(table, build, ids):
    if table in ["DNase", "TFBS"]:
        query = "SELECT r.regionID AS regionID, \
        t.chromosome AS {table}_chrom, \
        t.start AS {table}_start, \
        t.stop AS {table}_stop, \
        t.name AS {table}_name, \
        tt.cell_or_tissue AS {table}_tissue \
        FROM {build}_Regions AS r \
        INNER JOIN {build}_regionID_to_{table} AS link ON r.regionID = link.regionID \
        INNER JOIN {build}_{table} AS t ON link.link_ID = t.regionID \
        INNER JOIN {build}_{table}_tissues AS tt ON t.regionID = tt.regionID \
        WHERE r.regionID IN ({regions})".format(
            regions=','.join(['?']*len(ids)),
            build=build,
            table=table)
    elif table == "dbSuper":
        query = "SELECT r.regionID AS regionID, \
        t.chromosome AS {table}_chrom, \
        t.start AS {table}_start, \
        t.stop AS {table}_stop, \
        t.regionID AS {table}_name, \
        tt.cell_or_tissue AS {table}_tissue \
        FROM {build}_Regions AS r \
        INNER JOIN regionID_to_{table} AS link ON r.regionID = link.regionID \
        INNER JOIN {build}_{table} AS t ON link.link_ID = t.regionID \
        INNER JOIN {table}_tissues AS tt ON t.regionID = tt.regionID \
        WHERE r.regionID IN ({regions})".format(
            regions=','.join(['?']*len(ids)),
            build=build,
            table=table)
    elif table == "UCNE":
        query = "SELECT r.regionID AS regionID, \
        t.chromosome AS {table}_chrom, \
        t.start AS {table}_start, \
        t.stop AS {table}_stop, \
        t.regionID AS {table}_name \
        FROM {build}_Regions AS r \
        INNER JOIN regionID_to_{table} AS link ON r.regionID = link.regionID \
        INNER JOIN {build}_{table} AS t ON link.link_ID = t.regionID \
        WHERE r.regionID IN ({regions})".format(
            regions=','.join(['?']*len(ids)),
            build=build,
            table=table)
    elif table == "regions":
        query = "SELECT r.regionID, r.chromosome, r.start, r.stop, \
        r.type, r.std_type, r.DB_source, r.PhyloP100_median, r.constrain_pct, \
        GROUP_CONCAT(DISTINCT g.gene_symbol), \
        r.closestGene_symbol, r.closestGene_ensg, r.closestGene_dist, \
        GROUP_CONCAT(DISTINCT t.cell_or_tissue), \
        GROUP_CONCAT(DISTINCT m.method), \
        GROUP_CONCAT(DISTINCT p.phenotype) \
        FROM {build}_Regions AS r \
        LEFT JOIN tissues AS t ON r.regionID = t.regionID \
        LEFT JOIN genes AS g ON r.regionID = g.regionID \
        LEFT JOIN methods AS m ON r.regionID = m.regionID \
        LEFT JOIN phenos AS p ON r.regionID = p.regionID \
        WHERE r.regionID IN ({regions}) \
        GROUP BY r.regionID".format(
            regions=','.join(['?']*len(ids)),
            build=build)
    elif table == "gene_details":
        query = "SELECT r.regionID, r.chromosome, r.start, r.stop, r.std_type, \
        g.gene_symbol, m.method, t.cell_or_tissue \
        FROM {build}_Regions AS r \
        LEFT JOIN genes AS g ON r.regionID = g.regionID \
        LEFT JOIN tissues AS t ON g.interactionID = t.regionID \
        LEFT JOIN methods AS m ON g.interactionID = m.regionID \
        WHERE r.regionID IN ({regions})".format(
            regions=','.join(['?']*len(ids)),
            build=build)
    elif table == "pheno_details":
        query = "SELECT r.regionID, r.chromosome, r.start, r.stop, r.std_type, \
        p.phenotype, p.method, p.DB_source \
        FROM {build}_Regions AS r \
        LEFT JOIN phenos AS p ON r.regionID = p.regionID \
        WHERE r.regionID IN ({regions})".format(
            regions=','.join(['?']*len(ids)),
            build=build)
    return query

#get data from the database
def getData(conn, query, ids):
    ids = list(ids)
    cur = conn.cursor()
    cur.execute(query,ids)
    #print(query)
    #print(ids)
    results = cur.fetchall()
    if results:
        return results
    else:
        return False

#check file existance
def checkfile(files, operation="open", mode="stop", logger=None):
    if isinstance(files, str): files = [files]
    
    outfiles = []
    for myfile in files:
        if operation == "write":
            if os.path.isfile(myfile):
                if mode == "rename":
                    filebase, extension = os.path.splitext(myfile)
                    newfilename = filebase + '.' + now("") + extension
                    logger.warning("%s already exists", myfile)
                    logger.warning("Saving to: %s", newfilename)
                    outfiles.append(newfilename)
                elif mode == "overwrite":
                    logger.warning("%s already exists and will be overwritten!", myfile)
                    os.remove(myfile)
                    outfiles.append(myfile)	
                elif mode == "stop":
                    logger.critical("%s already exists! Exiting now!", myfile)
                    sys.exit()
            else:
                outfiles.append(myfile)
        elif operation == "open":
            if not os.path.isfile(myfile):
                logger.critical("Unable to open file %s", myfile)
                sys.exit()
    
    if len(outfiles) == 1: 
        return outfiles[0]
    else:
        return outfiles

#get results tables from db based on prepared query
def get_result_tables(db, build, region_ids, var_id=False, logger=None):
    result_tables = dict()
    for table in TABLES:            
        query = prepare_query(table, build, region_ids)
        records = getData(db,query,region_ids)
        if records:
            dataframe = pd.DataFrame.from_records(getData(db,query,region_ids))
            dataframe.columns = TAB_HEADERS[table]
            dataframe.iloc[:,2] = pd.to_numeric(dataframe.iloc[:,2])
            dataframe.iloc[:,3] = pd.to_numeric(dataframe.iloc[:,3])
            if var_id: dataframe['var_id'] = "VAR_ID"
            result_tables[table] = dataframe
        #else:
        #    result_tables[table] = pd.DataFrame
        if logger: logger.info("Query completed for %s table", table)
    return result_tables

#Process result tables add variant ids
def fillVarID(var_id, query_tables):
    var_tables = dict()
    for table in TABLES:
        if table in query_tables.keys():
            if not query_tables[table].empty: 
                df = query_tables[table]
                variant_fields = var_id.split("_")
                variant_chrom = variant_fields[0]
                variant_pos = int(variant_fields[1])
                df.loc[(df.iloc[:,1] == variant_chrom) & (df.iloc[:,2] <= variant_pos) & (df.iloc[:,3] >= variant_pos), 'var_id'] += "," + var_id
                var_tables[table] = df
    return var_tables

#Save tables to disk
def saveTables(tables,out_tables,logger):
    for table in tables:
        outfile = out_prefix + "." + table + ".tsv"
        if table in out_tables.keys():
            if not out_tables[table].empty:
                out_tables[table] = out_tables[table].drop_duplicates()
                out_tables[table].to_csv(outfile,sep="\t",index=False)
                logger.info("%s table saved to: %s", table, outfile)
            else:
                logger.info("No info found for %s table", table)
        else:
            logger.info("No info found for %s table", table)

###############
## Arguments ##
###############
parser = argparse.ArgumentParser(description='Script to annotate regulatory informations')
inputs = parser.add_mutually_exclusive_group(required=True)
inputs.add_argument("-v", "--vcf", help="Input vcf[.gz] file. Must be annotated with GREEN-VARAN", action="store")
inputs.add_argument("-r", "--regIDs", help="Comma separated list of region IDs or file with a list of region IDs", action="store")
inputs.add_argument("-t", "--table", help="Tab-separated file with col1: variant (chr_pos_ref_alt); col2: comma-separated list of region IDs", action="store")

parser.add_argument("-o", "--outprefix", help="Prefix for output files", action="store", required=True)

parser.add_argument("-b", "--build", help="Genome build", action="store", choices=['GRCh37', 'GRCh38'], required=True)
parser.add_argument("--regDB", help="RegDB database file", action="store", required=False, default=DB_FILE)
parser.add_argument("--logfile", help="Log file", action="store", required=False)
args = parser.parse_args()

###########################
## Logging configuration ##
###########################
if args.logfile:
    logfile = args.logfile
else:
    logfile = args.outprefix + "_" + now("","long") + ".log"

#config log for file
logging.basicConfig(filename=logfile, filemode='w', level=logging.INFO, \
    format='%(asctime)s %(name)-12s %(levelname)-8s %(message)s')
logger=logging.getLogger('NCANNO')

#config log for stdout
console = logging.StreamHandler()
console.setLevel(logging.INFO)
formatter = logging.Formatter('%(asctime)s: %(levelname)-8s %(message)s', datefmt = '%y-%m-%d %H:%M:%S')
console.setFormatter(formatter)
logger.addHandler(console)

#####################
### INITIAL SETUP ###
#####################
start_time = time.time()
regDB = args.regDB
build = args.build
out_prefix = args.outprefix
checkfile(regDB, logger=logger)

logger.info("#####  GREEN-DB QUERY version %s  #####", str(VERSION))
logger.info("# EFFECTIVE CONFIGURATION #")
logger.info("Python version: %s.%s", str(sys.version_info[0]), str(sys.version_info[1]))
if sys.version_info[0] < 3:
    logger.critical("Python version 3 is required")

#Detected configuration
logger.info("Genome build: %s", build)
logger.info("GREEN-DB: %s", regDB)

#Open connection to database
anno_db = create_connection(regDB)
if anno_db is None:
    logger.critical("Failed to connect to Regulatory regions db")
    sys.exit()

logger.info("# GREEN-DB QUERY START #")

### Query from region list without variants ###
if args.regIDs:
    region_IDs = readList(args.regIDs)
    logger.info("Region IDs input with %d IDs", len(region_IDs))
    out_tables = get_result_tables(anno_db, build, region_IDs, logger=logger)
    saveTables(TABLES,out_tables,logger)
    exitNow(start_time)

### Load var and regID from table ###
if args.table:
    checkfile(args.table, logger=logger) 
    logger.info("Variant table input: %s", args.table)
    region_IDs = set()
    variants = set()
    with open(args.table) as f:
        line = f.readline()
        while line:
            line = line.rstrip("\n")
            line = line.split("\t")
            
            variants.add(line[0])
            regions = set(line[1].split(","))
            if len(regions) > 0:
                region_IDs.update(regions)
            line = f.readline()
    logger.info("%d variants and %d regions read from file", len(variants), len(region_IDs))

### Load var and regID from VCF ###
if args.vcf:
    checkfile(args.vcf, logger=logger) 
    logger.info("VCF input: %s", args.vcf)

    logger.info("Reading variants and region IDs")
    region_IDs = set()
    variants = set()
    for variant in VCF(args.vcf):
        var_id = "_".join([str(x) for x in [variant.CHROM, variant.start, variant.REF, variant.ALT[0]]])
        nc_regid = variant.INFO.get("NC_regionID", False)
        nc_anno = variant.INFO.get("NC_ANNO", False)
        if nc_regid or nc_anno:
            variants.add(var_id)
        if nc_regid:
            region_IDs.update(nc_regid.split(","))
        elif nc_anno:
            nc_anno = nc_anno.split('|')
            region_IDs.update(nc_anno[1].split(","))
    
    logger.info("%d variants and %d regions read from file", len(variants), len(region_IDs))

### Query DB ###
step_time = time.time()
query_tables = get_result_tables(anno_db, build, region_IDs, var_id=True, logger=logger)
elapsed_time = time.time() - step_time
logger.info("Finished extracting regions from DB in %s sec", round(elapsed_time,3))

### Annotate var_id and prepare output tables ###
logger.info("Variant annotation started")
step_time = time.time()
interval_time = time.time()
n = 0
for v in variants:
    query_tables = fillVarID(v, query_tables)
    n = n + 1
    if n % 10 == 0:
        elapsed_time = time.time() - interval_time
        interval_time = time.time()
        logger.info("Annotated 10 variants in %s sec", round(elapsed_time,3))
elapsed_time = time.time() - step_time
logger.info("Finished variant annotation in %s sec", round(elapsed_time,3))

logger.info("Clean output tables")
for table in TABLES:
    if table in query_tables.keys():
        if not query_tables[table].empty:
            df = query_tables[table]
            df = df[df.var_id != "VAR_ID"]
            df['var_id'] = df['var_id'].str.replace('VAR_ID,', '')
            query_tables[table] = df

saveTables(TABLES,query_tables,logger)

exitNow(start_time)

#TODO Develop a plotting system

File: greendb_query.nim
import db_sqlite
import times
from math import floor
import tables
import logging
import argparse
import strformat
from os import createDir
from times import cpuTime

const
    GreendbVersions = @["v2.0", "v2.5"] 
    OUTTABS_HEADERS = {
        "regions" : @["regionID","chrom","start","stop","type","std_type","DB_source","detection_method","PhyloP100_median","constraint_pct","controlled_genes","closestGene_symbol", "closestGene_dist","closestProtGene_symbol","closestProtGene_dist","cell_or_tissues","phenotype"],
        "gene_details" : @["regionID","chrom","start","stop","std_type","controlled_gene","same_TAD","detection_method","tissue_of_interaction"],
        "pheno_details" : @["regionID","chrom","start","stop","std_type","phenotype","detection_method","DB_source"],
        "DNase" : @["regionID","DNase_chrom","DNase_start","DNase_stop","DNase_ID","DNase_cell_or_tissue"],
        "dbSuper" : @["regionID","dbSuper_chrom","dbSuper_start","dbSuper_stop","dbSuper_ID","dbSuper_cell_or_tissue"],
        "TFBS" : @["regionID","TFBS_chrom","TFBS_start","TFBS_stop","TF_name","TFBS_cell_or_tissue"],
        "UCNE" : @["regionID","UCNE_chrom","UCNE_start","UCNE_stop","UCNE_ID"]
    }.toTable
    STDCHROMS = @["chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9", 
        "chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19",
        "chr20","chr21","chr22","chrX","chrY","chrM"]

type
    Interval = object
        chrom: cstring
        start: int
        stop: int

proc timer(t0: var float): string {.discardable.} =
    let elapsed_time = cpuTime() - t0
    let m = floor(elapsed_time / 60)
    let s = elapsed_time - (m * 60)
    t0 = cpuTime()
    return fmt"{m} min {s:1.2f} sec"

proc initLogger(cl: var ConsoleLogger, fl: var FileLogger, log: string, outvcf: string) =
    var logfilename: string
    if log.len == 0:
        logfilename = parentDir(outvcf) & "/greenvaran_" & format(now().toTime, "yyyyMMdd'_'HHmmss", utc()) & ".log"
    else:
        logfilename = log
    cl = newConsoleLogger(fmtStr="[$datetime] - $levelname: ", levelThreshold=lvlInfo)
    fl = newFileLogger(logfilename, fmtStr="[$datetime] - $levelname: ", levelThreshold=lvlDebug)

proc getItems(s: string, class: string, nochr: bool): seq[string] =
    if s.len > 0:
        try:
            let f = open(s)
            var line: string
            defer: close(f)
            while f.read_line(line):
                result.add(line)
        except IOError:
            result = split(s, ",")
        
    if result.len == 0 and class == "chromosome":
        if nochr:
            for c in STDCHROMS: result.add(c.replace("chr", ""))
        else:
            result = STDCHROMS 

proc len(x: Interval): int {.inline.} =
    result = x.stop - x.start

proc overlap(x: Interval, y: Interval): (float, int) =
#Given an interval x return the fraction and n bases in x overlapped by y
    var 
        overlap_pct: float
        overlap_n: int
    if x.chrom != y.chrom:
        return (0.0, 0)
    else:
        #Note x is expected to be sorted before y
        if x.start <= y.start:
            overlap_n = x.stop - y.start
        else:
            overlap_n = y.stop - x.start
        
        overlap_pct = overlap_n / x.len

        return (overlap_pct, overlap_n)

proc makeinterval(s: seq[string], i: var Interval, idx: seq[int] = @[1,2,3]): bool =
    try:
        i.chrom = s[idx[0]].replace("chr", "")
        i.start = parseInt(s[idx[1]])
        i.stop = parseInt(s[idx[2]])
        if i.start > i.stop: 
            return false
        return true
    except:
        return false

proc prepare_query(table: string, build: string, ids: seq[string]): SqlQuery =
    #prepare query for the requested output table
    let regions = repeat("?", ids.len).join(",")
    var query: string
    case table:
        of "DNase", "TFBS":
            query = fmt"""SELECT r.regionID AS regionID, 
            t.chromosome AS {table}_chrom, 
            t.start AS {table}_start, 
            t.stop AS {table}_stop, 
            t.name AS {table}_name, 
            tt.cell_or_tissue AS {table}_tissue 
            FROM {build}_Regions AS r 
            INNER JOIN {build}_regionID_to_{table} AS link ON r.regionID = link.regionID 
            INNER JOIN {build}_{table} AS t ON link.link_ID = t.regionID 
            INNER JOIN {build}_{table}_tissues AS tt ON t.regionID = tt.regionID 
            WHERE r.regionID IN ({regions})"""
        of "dbSuper":
            query = fmt"""SELECT r.regionID AS regionID, 
            t.chromosome AS {table}_chrom, 
            t.start AS {table}_start, 
            t.stop AS {table}_stop, 
            t.regionID AS {table}_name, 
            tt.cell_or_tissue AS {table}_tissue 
            FROM {build}_Regions AS r 
            INNER JOIN regionID_to_{table} AS link ON r.regionID = link.regionID 
            INNER JOIN {build}_{table} AS t ON link.link_ID = t.regionID 
            INNER JOIN {table}_tissues AS tt ON t.regionID = tt.regionID 
            WHERE r.regionID IN ({regions})"""
        of "UCNE":
            query = fmt"""SELECT r.regionID AS regionID, 
            t.chromosome AS {table}_chrom, 
            t.start AS {table}_start, 
            t.stop AS {table}_stop, 
            t.regionID AS {table}_name 
            FROM {build}_Regions AS r 
            INNER JOIN regionID_to_{table} AS link ON r.regionID = link.regionID 
            INNER JOIN {build}_{table} AS t ON link.link_ID = t.regionID 
            WHERE r.regionID IN ({regions})"""
        of "regions":
            query = fmt"""SELECT r.regionID, r.chromosome, r.start, r.stop, 
            r.type, r.std_type, r.DB_source, GROUP_CONCAT(DISTINCT m.method), 
            r.PhyloP100_median, r.constrain_pct, 
            GROUP_CONCAT(DISTINCT g.gene_symbol), 
            r.closestGene_symbol, r.closestGene_dist, 
            r.closestProt_symbol, r.closestProt_dist,
            GROUP_CONCAT(DISTINCT t.cell_or_tissue), 
            GROUP_CONCAT(DISTINCT p.phenotype) 
            FROM {build}_Regions AS r 
            LEFT JOIN tissues AS t ON r.regionID = t.regionID 
            LEFT JOIN genes AS g ON r.regionID = g.regionID 
            LEFT JOIN methods AS m ON r.regionID = m.regionID 
            LEFT JOIN phenos AS p ON r.regionID = p.regionID 
            WHERE r.regionID IN ({regions}) 
            GROUP BY r.regionID"""
        of "gene_details":
            query = fmt"""SELECT r.regionID, r.chromosome, r.start, r.stop, r.std_type, 
            g.gene_symbol, g.sameTAD, m.method, t.cell_or_tissue 
            FROM {build}_Regions AS r 
            LEFT JOIN genes AS g ON r.regionID = g.regionID 
            LEFT JOIN tissues AS t ON g.interactionID = t.regionID 
            LEFT JOIN methods AS m ON g.interactionID = m.regionID 
            WHERE r.regionID IN ({regions})"""
        of "pheno_details":
            query = fmt"""SELECT r.regionID, r.chromosome, r.start, r.stop, r.std_type, 
            p.phenotype, p.method, p.DB_source 
            FROM {build}_Regions AS r 
            LEFT JOIN phenos AS p ON r.regionID = p.regionID 
            WHERE r.regionID IN ({regions})"""
    result = sql(query)

proc parse_variantID(v: string, varfields: var seq[string]): bool {.discardable.} =
    var 
        chrom: string
        start: string
        stop: string
    let fields = v.split("_")
    if fields.len == 4:
        #for smallvar format is chr_pos_ref_alt
        chrom = fields[0]
        start = fields[1]
        stop = fields[1]
        try: 
            discard parseInt(start)
            varfields = @[chrom, start, stop]
            return true
        except:
            return false
    elif fields.len == 5:
        #for sv format is chr_start_stop_ref_alt
        chrom = fields[0]
        start = fields[1]
        stop = fields[2]
        try: 
            discard parseInt(start)
            discard parseInt(stop)
            varfields = @[chrom, start, stop]
            return true
        except:
            return false
    else:
        return false

proc getRegions(db: DbConn, q: SqlQuery, v: string, r: var seq[string]): bool {.discardable.} =
    var varfields: seq[string]
    if not parse_variantID(v, varfields): 
        debug(fmt"Failed to parse variant {v}")
        return false

    for row in db.fastRows(q, varfields):
        r.add(row[0])
    return true

proc makeTables(db: DbConn, regions: seq[string], build: string, tabs: Table[string, seq[string]], outp: string, variant: string = ""): bool {.discardable.} =    
    var runok = true
    for table, header in tabs.pairs:
        let outfile = open(fmt"{outp}.{table}.tsv", fmWrite)
        if variant != "":
            outfile.write_line(header.join("\t") & "\tVariantID")
        else:
            outfile.write_line(header.join("\t"))
        
        let query = prepare_query(table, build, regions)
        for row in db.fastRows(query, regions):
            if variant != "":
                var 
                    rinterval: Interval
                    vinterval: Interval
                    varfields: seq[string]
                    varout: string
                if parse_variantID(variant, varfields):   
                    if makeinterval(row, rinterval) and makeinterval(varfields,vinterval, @[0,1,2]):
                        let (pctoverlap, noverlap) = vinterval.overlap(rinterval)
                        if noverlap > 0:
                            varout = variant
                        else:
                            varout = "NOVARIANT"
                    else:
                        debug(fmt"Error occurred parsing interval for region {row[0]} and variant {variant}")
                        varout = "NA"
                        runok = false
                else:
                    debug(fmt"Failed to parse variant {variant}")
                    varout = "NA"
                    runok = false  
                outfile.write_line(row.join("\t") & "\t" & varout)
            else:
                outfile.write_line(row.join("\t"))
        close(outfile)
    return runok

proc main* (dropfirst:bool=false) =
    var p = newParser("greendb_query"):
        option("-d", "--db", help="GREEN-DB SQlite file .db")
        option("-o", "--out", help="output prefix. A file prefix for --regions or a directory for --tab")
        option("-t", "--tab", help="input tab-separated file [col1:varid, col2:regiond_IDs]")
        option("-r", "--regids", help="list of region IDs. either comma-separated list of txt file")
        option("-v", "--variants", help="list of variants (chr_pos_ref_alt). either comma-separated list of txt file")
        option("-g", "--genome", help="genome build for query", choices = @["GRCh37","GRCh38"], default=some("GRCh38"))
        option("--log", help="Log file. Default is greenvaran_[now].log")

    try:
        let opts = p.parse() 
    except ShortCircuit as e:
        if e.flag == "argparse_help":
            echo p.help
            quit QuitSuccess
    except UsageError:
        stderr.writeLine getCurrentExceptionMsg() 
        echo "Use --help for usage information"
        quit QuitSuccess

    let opts = p.parse() 

    #Set logging
    var
        consoleLog: ConsoleLogger
        fileLog: FileLogger
    initLogger(consoleLog, fileLog, opts.log, opts.out)
    addHandler(consoleLog)
    addHandler(fileLog)

    #Start
    var start_time = cpuTime()
    info("=== GREEN-DB QUERY ===")
    info(fmt"Compatible with GREEN-DB ver: {$GreendbVersions}")
    info(fmt"Genome build: {opts.genome}")
    info(fmt"Open database from: {opts.db}")

    #Check mandatory arguments
    if opts.db.len == 0 or opts.genome.len == 0 or opts.out.len == 0:
        fatal("One of --db, --genome or --out is missing")
        quit "", QuitFailure

    #Check mode
    if opts.regids.len == 0 and opts.tab.len == 0 and opts.variants.len == 0:
        fatal("One of --regids, --tab or --variants must be provided")
        quit "", QuitFailure
    if opts.regids.len > 0 and opts.tab.len > 0 and opts.variants.len > 0:
        fatal("Only one of --regids, --tab or --variants can be used")
        quit "", QuitFailure
    var runmode: string
    if opts.regids.len > 0: runmode = "regions"
    if opts.tab.len > 0: runmode = "tab"
    if opts.variants.len > 0: runmode = "variants"

    #Open database
    let db = open(opts.db, "", "", "")
    db.exec(sql"PRAGMA synchronous = OFF")
    db.exec(sql"PRAGMA journal_mode = OFF")
    db.exec(sql"PRAGMA locking_mode = EXCLUSIVE")

    #Get regions
    var warning = false
    case runmode:
        of "regions":
            let regions = getItems(opts.regids,"regions", false)
            info(fmt"Regions in input: {regions.len}")
            makeTables(db, regions, opts.genome, OUTTABS_HEADERS, opts.out)
        of "tab":
            info(fmt"Reading from: {opts.tab}")
            var n = 0
            var nregions = 0
            for line in lines(opts.tab) :
                n += 1
                let fields = line.split("\t")
                createDir(fmt"{opts.out}/{fields[0]}")
                let outprefix = fmt"{opts.out}/{fields[0]}/{fields[0]}"
                let regions = fields[1].split(",")
                nregions += regions.len
                if not makeTables(db, regions, opts.genome, OUTTABS_HEADERS, outprefix, fields[0]):
                    warning = true
            info(fmt"{n} variants and {nregions} regions processed from input")
        of "variants":
            var nregions = 0
            let variants = getItems(opts.variants,"variants", false)
            info(fmt"Variants in input: {variants.len}")
            for v in variants:
                createDir(fmt"{opts.out}/{v}")
                let outprefix = fmt"{opts.out}/{v}/{v}"
                var regions: seq[string]
                let query = sql(fmt"""SELECT regionID FROM {opts.genome}_Regions 
                                        WHERE chromosome = ?
                                        AND start <= ?
                                        AND stop >= ?""")
                if not getRegions(db, query, v, regions): warning = true
                nregions += regions.len
                debug(fmt"{regions.len} regions extracted for variant {v}")
                if not makeTables(db, regions, opts.genome, OUTTABS_HEADERS, outprefix, v):
                    warning = true
            info(fmt"Total of {nregions} regions extracted for input variants")

    db.close()
    if warning: logging.error("Some regions of variants failed to parse, see log for more information")
    info(fmt"All done - Completed in {timer(starttime)}")
        
when isMainModule:
    main()

    

